Lib/zipimport.py

```diff
@@ -0,0 +1,640 @@
+        raise ZipImportError("can't decompress data; zlib not available")
+    finally:
+        _importing_zlib = False
+
+    _bootstrap._verbose_message('zipimport: zlib available')
+    return decompress
+
+# Given a path to a Zip file and a toc_entry, return the (uncompressed) data.
+def _get_data(archive, toc_entry):
+    datapath, compress, data_size, file_size, file_offset, time, date, crc = toc_entry
+    if data_size < 0:
+        raise ZipImportError('negative data size')
+
+    with _io.open(archive, 'rb') as fp:
+        # Check to make sure the local file header is correct
+        try:
+            fp.seek(file_offset)
+        except OSError:
+            raise ZipImportError(f"can't read Zip file: {archive!r}", path=archive)
+        buffer = fp.read(30)
+        if len(buffer) != 30:
+            raise EOFError('EOF read where not expected')
+
+        if buffer[:4] != b'PK\x03\x04':
+            # Bad: Local File Header
+            raise ZipImportError(f'bad local file header: {archive!r}', path=archive)
+
+        name_size = _unpack_uint16(buffer[26:28])
+        extra_size = _unpack_uint16(buffer[28:30])
+        header_size = 30 + name_size + extra_size
+        file_offset += header_size  # Start of file data
+        try:
+            fp.seek(file_offset)
+        except OSError:
+            raise ZipImportError(f"can't read Zip file: {archive!r}", path=archive)
+        try:
+            raw_data = fp.read(data_size)
+        except OSError:
+            raise OSError("zipimport: can't read data")
+        if len(raw_data) != data_size:
+            raise OSError("zipimport: can't read data")
+
+    if compress == 0:
+        # data is not compressed
+        return raw_data
+
+    # Decompress with zlib
+    try:
+        decompress = _get_decompress_func()
+    except:
+        raise ZipImportError("can't decompress data; zlib not available")
+    return decompress(raw_data, -15)
+
+
+# Lenient date/time comparison function. The precision of the mtime
+# in the archive is lower than the mtime stored in a .pyc: we
+# must allow a difference of at most one second.
+def _eq_mtime(t1, t2):
+    # dostime only stores even seconds, so be lenient
+    return abs(t1 - t2) <= 1
+
+# Given the contents of a .py[co] file, unmarshal the data
+# and return the code object. Return None if it the magic word doesn't
+# match (we do this instead of raising an exception as we fall back
+# to .py if available and we don't want to mask other errors).
+def _unmarshal_code(pathname, data, mtime):
+    if len(data) < 16:
+        raise ZipImportError('bad pyc data')
+
+    if data[:4] != _bootstrap_external.MAGIC_NUMBER:
+        _bootstrap._verbose_message('{!r} has bad magic', pathname)
+        return None  # signal caller to try alternative
+
+    flags = _unpack_uint32(data[4:8])
+    if flags != 0:
+        # Hash-based pyc. We currently refuse to handle checked hash-based
```

Comment:
I'm not so sure about this.  I don't have a good intuition either way.  Maybe @benjaminp (PEP 552 author) has an opinion?

Can you also add a reference to PEP 552?

Reply:
This is what the current `zipimport` implementation does, too. We may want to change the behavior in the future, but it doesn't need to happen in this rewrite.

Reply:
Thanks @benjaminp - @serhiy-storchaka as with the general cleanup, I think it's cool if you want to defer it, but if so, can you open an issue so it doesn't get lost (even if it takes a while to implement)?

Reply:
@serhiy-storchaka Doing the clean up in a follow on branch is fine.